strawberry/types/generics.py

```diff
@@ -1,35 +1,55 @@
 import builtins
 import dataclasses
-from typing import Dict, Iterable, Tuple, Type, cast
+from typing import Dict, Iterable, Tuple, Type, Union, cast
 
 from strawberry.union import StrawberryUnion, union
 from strawberry.utils.str_converters import capitalize_first
-from strawberry.utils.typing import is_type_var
+from strawberry.utils.typing import is_type_var, is_union
 
 from .types import FederationFieldParams, FieldDefinition, TypeDefinition
 
 
-def get_name_from_types(types: Iterable[Type]):
-    return "".join([capitalize_first(type.__name__) for type in types])
+def get_name_from_types(types: Iterable[Union[Type, StrawberryUnion]]):
+    names = []
+
+    for type_ in types:
+        if isinstance(type_, StrawberryUnion):
+            return type_.name
+        else:
+            name = capitalize_first(type_.__name__)
```

Comment:
This is quite ugly, but I'm pretty sure that we'll improve this when the objectification proposed by @BryceBeagle gets done :)